### PR TITLE
libvmi: introduce WithRunStrategy and prefer it over WithRunning in many tests

### PR DESCRIPTION
**Pull request by akalenyu**
### What this PR does
As we have a proposal to deprecate the running field (https://github.com/kubevirt/kubevirt/issues/11993)
Let's have some tests use run strategies instead of the running field via libvmi,
similarly to how we would also steer users towards the same direction.

Note we cannot simply abandon running (semver) and thus we will keep it in many test suites.
Storage tests were picked for some coverage of WithRunStrategy 
since they're more of a "casual" user of the running/runstrategy API
(indifferent to the specifics of running vs runstrategy)

For more info about the reasoning for runstrategies:
https://kubevirt.io/user-guide/compute/run_strategies/

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/libvmi/vm.go
Comment (iholder101) [inline pkg/libvmi/vm.go:66]: WDYT about also assigning `vm.Spec.Running = nil` here?
I think it makes sense as it would help get rid of "both .spec.running and .spec.runStrategy are set" error.

**Review by iholder101 — COMMENTED** [review]
Thank you @akalenyu! Another step towards the right direction! :tada:

File: pkg/libvmi/vm.go
Comment (akalenyu) [reply pkg/libvmi/vm.go:66]: I can try but what's happening in the tests is that many of them are setting running=true pre-creation https://github.com/kubevirt/kubevirt/blob/bee5cc119bfb95e6b41860da56c2ba2bbcbe28f3/tests/storage/snapshot.go#L336-L339

**Comment by akalenyu** [comment]
just pushed a rebase

**Comment by iholder101** [comment]
Hey @akalenyu!
Bear in mind that a similar PR is ongoing: https://github.com/kubevirt/kubevirt/pull/12029. For awareness to avoid you two working on the same code.

File: pkg/libvmi/vm.go
Comment (EdDev) [inline pkg/libvmi/vm.go:72]: It is better not to have this, but if this is requiring too many changes on other tests just mention it to and mention it is an exception.

Comment (akalenyu) [reply pkg/libvmi/vm.go:72]: Right, so it's just that there are too many WithRunning invocations
```bash
$ (cd tests; git grep "WithRunning" | wc -l)
29
```
and I think we want to keep them for the reason that we can't simply abandon Running, have to ensure it doesn't break.
I have a longer version of this explanation in the commit message:
https://github.com/kubevirt/kubevirt/pull/12023/commits/30a0671593813ae5f8b005227fe35938d323ebfd

WDYT about putting a comment on WithRunning that discourages it's use, along with the hidden detail of setting runStrategy to nil?

Comment (akalenyu) [reply pkg/libvmi/vm.go:72]: Or, we could also have a hacky WithRunning variation outside of libvmi, something like WithDiscouragedRunning under libvmifact

Comment (akalenyu) [reply pkg/libvmi/vm.go:76]: done, let me know what you think

Comment (EdDev) [reply pkg/libvmi/vm.go:76]: Please squash that last commit into the one that introduced this change.

Comment (akalenyu) [reply pkg/libvmi/vm.go:76]: done

File: tests/clone_test.go
Comment (EdDev) [inline tests/clone_test.go:193]: This is introducing new helpers, so I would suggest to:
- Implement them as regular functions and not closures.
- Consider placing them in a different common place (`libclientset`?). Maybe in the same package the current kubevirt `Client()` reside.
- Consider returning the error over asserting it inside the helper. It is hard to estimate from what stack level it is called, making it challenging to trace it well. Usually an `offset` is set for the assertions to solve that.
- Use `Patch` to avoid retry.

If you change these to a regular function and take care to return the error, we can do the rest as a follow up.

Comment (akalenyu) [reply tests/clone_test.go:193]: So these have no business being in common packages right now but I tried to explain the issue with the existing helpers in the commit message - https://github.com/kubevirt/kubevirt/pull/12023/commits/2654f2c5b628015c4171bd4478593086a5264615.
Maybe we tackle Start/StopVirtualMachine in a follow up since I imagine it's not going to be straightforward.

> Use Patch to avoid retry

Yes that would be ideal but unfortunately the error coming back from patch also needs to be asserted on

**Comment by akalenyu** [comment]
/test pull-kubevirt-e2e-k8s-1.30-sig-compute-migrations
unrelated

File: tests/clone_test.go
Comment (akalenyu) [reply tests/clone_test.go:12]: there's a `"k8s.io/utils/pointer"` import in the same file

File: tests/clone_test.go
Comment (EdDev) [reply tests/clone_test.go:193]: > > Use Patch to avoid retry
> 
> Yes that would be ideal but unfortunately the error coming back from patch also needs to be asserted on

The error needs to be asserted, sure. But it does not need to retried.

> Maybe we tackle Start/StopVirtualMachine in a follow up since I imagine it's not going to be straightforward.

I am fine with that, this is why I mentioned:
> If you change these to a regular function and take care to return the error, we can do the rest as a follow up.

Is this something you can take care here?

File: tests/clone_test.go
Comment (EdDev) [reply tests/clone_test.go:12]: > there's a "k8s.io/utils/pointer" import in the same file

Ohh, I did not notice that.

Then it is fine to do it as a follow up from my side.

Comment (akalenyu) [reply tests/clone_test.go]: Now using our pointer package :+1:

File: tests/clone_test.go
Comment (akalenyu) [reply tests/clone_test.go]: > Is this something you can take care here?

Sure - done

File: tests/clone_test.go
Comment (EdDev) [inline tests/clone_test.go:862]: - You could return directly here.
- You could avoid passing virt-client (as it is a singleton).

File: tests/clone_test.go
Comment (EdDev) [inline tests/clone_test.go:876]: - You could return directly here.
- You could avoid passing virt-client (as it is a singleton).

File: tests/clone_test.go
Comment (EdDev) [inline tests/clone_test.go:857]: I think it is safer and better to use `WithAdd`.

File: tests/clone_test.go
Comment (EdDev) [inline tests/clone_test.go:871]: I think it is safer and better to use `WithAdd`.

**Review by EdDev — APPROVED** [review]
Thanks.

I had a few comments inline and it will be great if you take care of them, but they are not blockers.

Thank you for the `libvmi` changes.
/approve

**Comment by EdDev** [comment]
Interesting, I did not know I got promoted to approve other SIGs changes.

Anyway, placing a hold so @akalenyu can decide if he wants to take care of the comments.
/hold

**Comment by dhiller** [comment]
> Interesting, I did not know I got promoted to approve other SIGs changes.
> 
> Anyway, placing a hold so @akalenyu can decide if he wants to take care of the comments. Please unhold when you think it is ready.
> 
> /hold

I see that you are an approver for:
* [.golangci.yml](https://github.com/kubevirt/kubevirt/blob/f124e9df70147191a3bab7b770e6d9659968d95b/OWNERS#L28)
* [pkg/libvmi](https://github.com/kubevirt/kubevirt/blob/main/pkg/libvmi/OWNERS#L3)
* [tests](https://github.com/kubevirt/kubevirt/blob/main/tests/OWNERS#L5)

So everything seems to be working as it should.

@EdDev do I overlook something here or why do you think you might not be eligible for approval?

File: tests/clone_test.go
Comment (akalenyu) [reply tests/clone_test.go:857]: Are you sure? I think only replace would work here as /spec/runStrategy exists at that point

File: tests/clone_test.go
Comment (akalenyu) [reply tests/clone_test.go:871]: Are you sure? I think only replace would work here as /spec/runStrategy exists at that point

Comment (EdDev) [reply tests/clone_test.go:871]: Yes, I'm sure. `add` will override or add any existing value.
It is unsafe to use `replace` because you require it to be there, and that is assured only if it was added in the first place and not removed after (e.g. an update could have removed it).

In general, I would avoid using `replace` as it has no added value and raises questions like the above.

Comment (akalenyu) [reply tests/clone_test.go]: Right. Addressed with some of the other comments
https://github.com/kubevirt/kubevirt/compare/65e089752d2e4db39fe6fc325dc2fba383c46cf0..961622903fd0e446efe859b11c654725e01a588d

File: tests/clone_test.go
Comment (akalenyu) [reply tests/clone_test.go]: Thanks. I honestly prefer passing the one from beforeeach instead of `virtClient := kubevirt.Client()`
but let me know if you insist on that one

Comment (EdDev) [reply tests/clone_test.go]: I thought we agreed passing all over the clientset is not useful and causes unneeded noise.
But I have no intention to insist on such nits given the context of this change.

**Review by EdDev — APPROVED** [review]

**Pull request merged**
